architecture/README.md

```diff
@@ -0,0 +1,190 @@
+# Architecture
+Adds a Lifecycle aware ViewModel solution with support for Navigation and Observables. Kaluga favours an MVVM based architecture due to its clean delegation between UI and View State.
+
+## ViewModels
+ViewModels can be created by subclassing the `BaseViewModel` class. The ViewModel runs within its own coroutine scope that exists for the entire lifetime of the ViewModel.
+In addition ViewModels have their own lifecycle, allowing them to the paused and resumed by the view. When the ViewModel is resumed a `CoroutineScope` with a lifecycle limited to the resumed state is provided.
+Launch any coroutines within this scope to automatically cancel when the ViewModel is paused.
+
+```kotlin
+class SomeViewModel : BaseViewModel() {
+
+    init {
+        coroutineScope.launch {
+            // Do stuff that happens even when viewModel paused
+        }
+    }
+
+    override fun onResume(scope: CoroutineScope) {
+        super.onResume(scope)
+
+        scope.launch {
+            // Do stuff that happens only when viewmodel is active
+        } 
+    }
+
+}
+```
+
+On Android the AndroidX `ViewModel` serves as a base. The Viewmodel lifecycle matches onResume/onPause of the `Activity` or `Fragment` that created it.
+To achieve this, the View should be bound using the `KalugaViewModelLifecycleObserver.bind()` method.
+For convenience default implementations for `Activity`, `Fragment`, and `DialogFragment` exist (`KalugaViewModelActivity`, `KalugaViewModelFragment`, and `KalugaViewModelDialogFragment` respectively).
+Alternatively the user can call `didResume()` and `didPause()` on the ViewModel manually in the `onResume()` and `onPause()` methods.
+
+On iOS automatic setup was omitted due to ObjC classes requiring to be final.
+To use it, the user must manually call the ViewModels `didResume`/`didPause` in `viewDidAppear`/`viewDidDisappear`.
+
+## Observables
+Kaluga supports `Observables` (one way binding) and `Subjects` (Two way binding). An Object can be created through a `ReadOnlyProperty` (making it immutable on both sides), a `Flow` (allowing the flow to modify the observer), or a `BaseFlowable` (allowing both the Flow and the owner of BaseFlowable to modify the observer.
+Subjects can be created using either a `ObservableProperty` or `BaseFlowable`. All these can easily be converted using `asObservable()` or `asSubject()` respectively.
+Observable values can be accessed through delegation. To account for the difference between empty values and optional values a `ObservableResult` containing either the value or a `Nothing` type is returned.
```

Comment:
> To account for the difference between empty values and optional values

What would be a meaning of an empty value permitted through the observable?
Could you provide an example where it would be useful?

In Swift/RxSwift whenever you're just interested that something has happened and you don't care about the value `Observable<Void>` is used. In my opinion `Observable<Optional<Type>>` rarely a good pattern.

Reply:
I've rephrased 'empty' to 'uninitialized'.

An observable can be created without having an initial value. At the same time, observing an optional value may be a valid case (e.g. single use events will need to set their data to 'null' after consumption). You can thus have both uninitialized non-optional observables and initialized optional observables. As such null it not a good indicator for initialization.

This Result class makes this distinction possible

Reply:
> An observable can be created without having an initial value.
This sounds exactly like a use case for a stateless observable.

Personally I'd prefer to have an option to just use a stateless observable instead of using such "workarounds".

Observing events (as opposed to observing values/states) is for me main use case for stateless observables. In this case you:
1. start with no initial value
2. you don't want a new subscriber to get "the last event" as subscriber is interested only in the events that will happen after it's a subscription

Reply:
Statelessness it not required here since Observables are purely used for data binding. All other cases should be handled through the kotlin provided Flow/Channel classes. Since Flows can be easily converted to/from Observables, this is interoperable.